proof/refine/ARM/Ipc_R.thy

```diff
@@ -3156,6 +3156,461 @@ lemma maybeReturnSc_corres:
+    apply (clarsimp simp add: caps_of_state_after_update cte_wp_at_after_update
+                              swp_def fun_upd_def obj_at_def)
+    apply (subst conj_assoc[symmetric])
+    apply (rule conjI[rotated])
+     apply (clarsimp simp add: ghost_relation_def)
+     apply (erule_tac x=ptr in allE)+
+     apply (clarsimp simp: obj_at_def a_type_def
+                     split: Structures_A.kernel_object.splits if_split_asm)
+    apply (fold fun_upd_def)
+    apply (simp only: pspace_relation_def simp_thms
+                      pspace_dom_update[where x="kernel_object.Reply _"
+                                          and v="kernel_object.Reply _",
+                                        simplified a_type_def, simplified])
+    apply (simp only: dom_fun_upd2 simp_thms)
+    apply (elim conjE)
+    apply (frule bspec, erule domI)
+    apply (rule conjI)
+     apply (rule ballI, drule(1) bspec)
+     apply (drule domD)
+     apply (clarsimp simp: project_inject split: if_split_asm kernel_object.split_asm)
+     apply (rename_tac bb aa ba)
+     apply (drule_tac x="(aa, ba)" in bspec, simp)
+     apply clarsimp
+     apply (frule_tac ko'="kernel_object.Reply reply" and x'=ptr in obj_relation_cut_same_type)
+        apply simp+
+     apply clarsimp
+      (* sc_replies_relation *)
+    apply (simp add: sc_replies_relation_def)
+    apply (clarsimp simp: sc_replies_of_scs_def map_project_def scs_of_kh_def)
+    apply (rule conjI; clarsimp simp: sc_of_def split: Structures_A.kernel_object.split_asm if_split_asm)
+    apply (drule_tac x=p in spec)
+   apply (subgoal_tac "((scs_of' b)(ptr := sc_of' (KOReply ae')) |> scReply) p = scReplies_of b p")
+     apply simp
+    apply (subgoal_tac "heap_list (replyPrevs_of b) (scReplies_of b p) (sc_replies z)")
+    apply (erule heap_list_heap_update_noop)
+    apply (case_tac "sc_replies z"; simp)
+     apply (clarsimp simp: sc_at_pred_n_def obj_at_def)
+    apply (drule_tac x=p in spec)
+    apply (simp add: typ_at'_def ko_wp_at'_def obj_at'_def project_inject opt_map_def sc_of_def)
+    apply (simp add: typ_at'_def ko_wp_at'_def obj_at'_def project_inject opt_map_def sc_of_def)
+  done
+  qed
+
+lemma setReply_corres_reply_not_queued:
+  "reply_relation r reply
+  \<Longrightarrow> corres dc (\<lambda>s. \<forall>p. sc_replies_sc_at (\<lambda>xs. rptr \<notin> set xs) p s) (reply_at' rptr)
+        (set_reply rptr r) (setReply rptr (replyPrev_update (\<lambda>_. scReplyOpt) reply))"
+  apply (rule corres_guard_imp)
+  apply (rule set_reply_corres)
+  apply (clarsimp simp: reply_relation_def )+
+  done
+
+lemma update_sc_reply_stack_update_ko_at'_corres:
+  "corres dc
+     (sc_at ptr and pspace_aligned and pspace_distinct)
+     (ko_at' sc' ptr and (\<lambda>s. heap_list (replyPrevs_of s) X y))
+     (update_sched_context ptr (sc_replies_update (\<lambda>_. y)))
+     (setSchedContext ptr (scReply_update (\<lambda>_. X) sc'))"
```

Comment:
nitpick: more descriptive variable names instead of `X` and `y`. Maybe `reply_ptr` and `replies`?